shiny/render/_data_frame_utils/_selection.py

```diff
@@ -0,0 +1,450 @@
+x = {"rows": [row_idx, row_idx + 3, row_idx + 5]}
+x = {"rows": [row_idx, row_idx + 3, row_idx + 5], "cols": None}
+
+# (same for cols)
+# select single col; all rows
+x = {"cols": col_idx}
+x = {"cols": col_idx, "rows": None}
+x = {"cols": [col_idx]}
+x = {"cols": [col_idx], "rows": None}
+# select multiple cols; all rows
+x = {"cols": [col_idx, col_idx + 3, col_idx + 5]}
+x = {"cols": [col_idx, col_idx + 3, col_idx + 5], "rows": None}
+
+
+class CellSelectionCell(TypedDict):
+    rows: int
+    cols: int
+
+
+class CellSelectionRegion(TypedDict):
+    rows: int | Sequence[int]
+    cols: int | Sequence[int]
+
+
+class CellSelectionRow(TypedDict):
+    rows: int | Sequence[int]
+    cols: NotRequired[None]
+
+
+class CellSelectionCol(TypedDict):
+    rows: NotRequired[None]
+    cols: int | tuple[int, ...]
+
+
+class CellSelectionNone(TypedDict):
+    rows: None
+    cols: None
+
+
+# class CellSelectionCore(TypedDict):
+#     rows: Sequence[int] | None
+#     cols: Sequence[int] | None
+# CellSelectionBase = Union[
+#     # SelectionLocationDict,
+#     CellSelectionCore,
+#     Literal["all", "none"],
+# ]
+
+
+# class CellSelectionAll(TypedDict):
+#     rows: True
+#     cols: True
+
+
+# For users to type in python
+CellSelectionDict = Union[
+    CellSelectionCell,
+    CellSelectionRegion,
+    CellSelectionRow,
+    CellSelectionCol,
+    CellSelectionNone,
+    # CellSelectionAll,
+]
+
+# @@@@@@@@@@@@@@
+
+
+# class SelectionLocationRegion(TypedDict):
+#     rows: LocationRange
+#     cols: LocationRange
+
+
+# class SelectionLocationRow(TypedDict):
+#     rows: tuple[int, ...]
+#     cols: Literal["all"]
+
+
+# class SelectionLocationCol(TypedDict):
+#     rows: Literal["all"]
+#     cols: LocationRange
+
+
+# class SelectionLocationAll(TypedDict):
+#     rows: Literal["all"]
+#     cols: Literal["all"]
+
+
+# SelectionLocationDict = Union[
+#     SelectionLocationRegion,
+#     SelectionLocationRow,
+#     SelectionLocationCol,
+#     SelectionLocationAll,
+# ]
+
+# More user friendly Selection type:
+CellSelection = Union[
+    # SelectionLocationDict,
+    CellSelectionDict,
+    Literal["all", "none"],
+]
+
+
+# Proposal, remove BrowserCellSelectionAll and BrowserCellSelectionNone as they can be represented by an empty Row/Col/Region
+class BrowserCellSelectionAll(TypedDict):
+    type: Literal["all"]
+
+
+class BrowserCellSelectionNone(TypedDict):
+    type: Literal["none"]
+
+
+class BrowserCellSelectionRow(TypedDict):
+    type: Literal["row"]
+    rows: tuple[int, ...]
+
+
+class BrowserCellSelectionCol(TypedDict):
+    type: Literal["col"]
+    cols: tuple[int, ...]
+
+
+class BrowserCellSelectionRegion(TypedDict):
+    type: Literal["region"]
+    rows: tuple[int, int]
+    cols: tuple[int, int]
+
+
+# For receiving selection info from JS:
+BrowserCellSelection = Union[
+    BrowserCellSelectionRegion,
+    BrowserCellSelectionRow,
+    BrowserCellSelectionCol,
+    BrowserCellSelectionAll,
+    BrowserCellSelectionNone,
+]
```

Comment:
I mentioned this elsewhere, but how will we support multiple types of selections in the same data table?